Fix floats in a full-bleed world

jasmussen opened the pull request:
This PR seeks to address #784 by applying an invisible and `calc`ulated outer margin to floats. This margin just happens to match the space surrounding the 700px max-width of the visual editor, making it look right.

I realize the calc looks advanced, but it really simplifies the responsive CSS.

jasmussen committed:
Fix problems with floated images & full-width

Fixes #784.

jasmussen committed:
Fix responsive.

youknowriad commented on blocks/library/image/style.scss:
```diff
@@ -9,11 +9,19 @@
+
+		@include break-medium() {
+			margin-left: calc( 50% - #{ $visual-editor-max-width / 2 } );
+		}
```
Aside: Do we expect block authors to implement this CSS styles (which are not that easy) in each floated block

jasmussen replied:
Good aside. And no, I've been thinking of it this way:

- Full-width can be implemented in many different ways, we've chosen one of many ways
- Theme authors should probably opt-in to the feature, probably something like `add_theme_support( 'full-width-images' );`
- When the theme has opted in, images simply get `align-full-width` as a CSS class. Or something in that vein.

youknowriad replied:
I'm asking about block authors too. I mean, should we move these styles outside the `image` block stylesheet and provide a way to reuse it across different blocks.

jasmussen replied:
Ah yes, good point. Taking a look.

jasmussen replied:
Should we move the visual styles out as part of this PR, or as part of for example #729? Tackling that one seems like a good successor to this one, as video should also have full width as an option.

youknowriad reviewed (commented):
I don't know if it's related but I'm still seeing an issue with floated blocks if they precede an embed block

youknowriad commented:
Here is the undesired behaviour

<img width="671" alt="screen shot 2017-05-16 at 10 34 09" src="https://cloud.githubusercontent.com/assets/272444/26099772/b776865a-3a23-11e7-9963-358b7c7c1fb0.png">

jasmussen commented:
Thanks for the screenshot, taking a look!

jasmussen committed:
Add float clearing to embed block.

This seems like the right thing to do, because videos normally have an intrinsice width and height, and so they would clear floats naturally just like how an image following a floating would. But because we are using CSS trickery to make the video responsive, we are essentially pulling the video out of the flow, which means it's technically abs positioned. By adding a clear, we emulate the behavior. The edgecases where the clear for an embed would be undesired are arguably edgecases.

jasmussen commented:
Upon discussing this with @mtias I've decided to close this branch, go back to the max-width, then take a pass at using JS for full-width images.
